### ISSUE TITLE
crash on add missing for tar.zst archives

### ISSUE THREAD

**Issue by pmiess** [0install/0publish-win#40]
if a tar.zst fie is entered and the add missing button pushed a 0publish crashes

```
System.TypeInitializationException: The type initializer for 'PerTypeValues`1' threw an exception. ---> System.IO.FileLoadException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at System.SpanHelpers.PerTypeValues`1.MeasureArrayAdjustment()
   at System.SpanHelpers.PerTypeValues`1..cctor()
   --- End of inner exception stack trace ---
   at ImpromptuNinjas.ZStd.ZStdDecompressStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at ICSharpCode.SharpZipLib.Tar.TarBuffer.ReadRecord()
   at ICSharpCode.SharpZipLib.Tar.TarBuffer.ReadBlock()
   at ICSharpCode.SharpZipLib.Tar.TarInputStream.GetNextEntry()
   at ZeroInstall.Store.Implementations.Archives.TarExtractor.ExtractArchive()
   at ZeroInstall.Store.Implementations.Archives.ArchiveExtractor.Execute()
   at NanoByte.Common.Tasks.TaskBase.Run(CancellationToken cancellationToken, ICredentialProvider credentialProvider, IProgress`1 progress)
   at NanoByte.Common.Tasks.TaskRunDialog.RunTask()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NanoByte.Common.ExceptionUtils.Rethrow(Exception exception)
   at ZeroInstall.Store.Implementations.Build.RecipeUtils.Apply(Archive step, String localPath, TemporaryDirectory workingDir, ITaskHandler handler, Object tag)
   at ZeroInstall.Publish.RetrievalMethodUtils.DownloadAndApply(DownloadRetrievalMethod retrievalMethod, ITaskHandler handler, ICommandExecutor executor)
   at ZeroInstall.Publish.RetrievalMethodUtils.DownloadAndApply(RetrievalMethod retrievalMethod, ITaskHandler handler, ICommandExecutor executor)
   at ZeroInstall.Publish.WinForms.Controls.RetrievalMethodEditor`1.CheckDigest(ITaskHandler handler, ICommandExecutor executor)
   at ZeroInstall.Publish.WinForms.Controls.RetrievalMethodEditor`1.buttonAddMissing_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Form.ShowDialog(IWin32Window owner)
   at ZeroInstall.Publish.WinForms.WelcomeForm.SwitchToMain(FeedEditing feedEditing)
   at ZeroInstall.Publish.WinForms.WelcomeForm.buttonOpen_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Forms.UnsafeNativeMethods.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.System.Windows.Forms.UnsafeNativeMethods.IMsoComponentManager.FPushMessageLoop(IntPtr dwComponentID, Int32 reason, Int32 pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Int32 reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Int32 reason, ApplicationContext context)
   at ZeroInstall.Publish.WinForms.Program.Main(String[] args)

System.IO.FileLoadException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
File name: 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.SpanHelpers.PerTypeValues`1.MeasureArrayAdjustment()
   at System.SpanHelpers.PerTypeValues`1..cctor()

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
```

the file
```xml
<?xml version="1.0" encoding="utf-8"?>
<interface xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://zero-install.sourceforge.net/2004/injector/interface http://0install.de/schema/injector/interface/interface.xsd http://0install.de/schema/desktop-integration/capabilities http://0install.de/schema/desktop-integration/capabilities/capabilities.xsd" uri="https://apps.0install.net/lib/gmp.xml" xmlns="http://zero-install.sourceforge.net/2004/injector/interface">
  <name>gmp</name>
  <summary xml:lang="en">A free library for arbitrary precision arithmetic (mingw-w64)</summary>
  <description xml:lang="en" />
  <icon href="https://raw.githubusercontent.com/0install/apps/master/lib/gmp.png" type="image/png" />
  <icon href="https://raw.githubusercontent.com/0install/apps/master/lib/gmp.ico" type="image/vnd.microsoft.icon" />
  <homepage>https://gmplib.org/</homepage>
  <package-implementation package="gmp" />
  <package-implementation package="dev-libs/gmp" distributions="Gentoo" />
  <package-implementation package="math/gmp" distributions="Ports" />
  <group license="LGPL3, GPL">
    <group version="6.2.1-1" released="2021-01-05">
      <implementation arch="Windows-i686" id="sha1new=202127badbbf73020e047b7cd746272fa898c22e">
        <manifest-digest sha256new="2QKO666TTOCECXIRGACE26BOTQIUTLUEFSXB3EQLUCJWXPU6LA3A" />
        <archive href="http://repo.msys2.org/mingw/i686/mingw-w64-i686-gmp-6.2.1-1-any.pkg.tar.zst" size="482992" type="application/x-zstd-compressed-tar" extract="mingw32" />
      </implementation>
      <implementation arch="Windows-x86_64" id="sha1new=18cca1453ce1038965e94c665c879164b0423021">
        <manifest-digest sha256new="BXZVDF3DOC4Y44YRPJDUILYVIBVTTHO5TEZE4RHUKMX3RYAED6KQ" />
        <archive href="http://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-gmp-6.2-1-1-any.pkg.tar.zst" size="497700" type="application/x-zstd-compressed-tar" extract="mingw64" />
      </implementation>
    </group>
  </group>
</interface>
```

**Comment by bastianeicher** [comment]
The crash you experienced was due to a missing assembly binding redirect. I've fixed this in 8e22306e3fd74eefea6c7bf03bcbafffc3ad6354. I'll update this issue when there's a new release.

The tar.zst active type is in fact listed in the Zero Install Feed Editor as the MIME type `application/x-zstd-compressed-tar`.

**Comment by pmiess** [comment]
Your right I missed it some how, sorry about that

**Comment by bastianeicher** [comment]
Fixed in [v2.19.0](https://github.com/0install/0publish-win/releases/tag/2.19.0).
